src/actions/channel_categories.ts

```diff
@@ -51,94 +50,66 @@ export function collapseCategory(categoryId: string) {
-            }
-
-            // Assume category IDs match up with the ones that are generated in the reducer
-            let categoryId: string;
-            if (isFavoriteChannel(myPreferences, channel.id)) {
-                categoryId = `${teamId}-favorites`;
-            } else if (channel.type === General.DM_CHANNEL || channel.type === General.GM_CHANNEL) {
-                categoryId = `${teamId}-direct_messages`;
-            } else {
-                categoryId = `${teamId}-channels`;
-            }
+        return dispatch({
+            type: ChannelCategoryTypes.RECEIVED_CATEGORY,
+            data: updatedCategory,
+        });
+    };
+}
 
-            channelsByCategory[categoryId].push(channel);
+export function fetchMyCategories(teamId: string) {
+    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
+        const currentUserId = getCurrentUserId(getState());
+
+        let data: OrderedChannelCategories;
+        try {
+            data = await Client4.getChannelCategories(currentUserId, teamId);
+        } catch (error) {
+            forceLogoutIfNecessary(error, dispatch, getState);
+            dispatch(logError(error));
+            return {error};
         }
 
-        for (const category of categories) {
-            let channels = channelsByCategory[category.id];
-
-            // Assume we've already loaded everything needed to sort the channels even though we likely won't have
-            // users loaded for DMs and GMs
-            if (channels.some((channel) => channel.type === General.DM_CHANNEL || channel.type === General.GM_CHANNEL)) {
-                channels = makeSortChannelsByNameWithDMs()(state, channels);
-            } else {
-                channels = makeSortChannelsByName()(state, channels);
-            }
-
-            dispatch({
-                type: ChannelCategoryTypes.RECEIVED_CATEGORY,
+        return dispatch(batchActions([
+            {
+                type: ChannelCategoryTypes.RECEIVED_CATEGORIES,
+                data: data.categories,
+            },
+            {
+                type: ChannelCategoryTypes.RECEIVED_CATEGORY_ORDER,
                 data: {
-                    ...category,
-                    channel_ids: channels.map((channel) => channel.id),
+                    teamId,
+                    order: data.order,
                 },
-            });
-        }
-
-        return {data: true};
+            },
+        ]));
     };
 }
 
 // addChannelToInitialCategory returns an action that can be dispatched to add a newly-joined or newly-created channel
 // to its either the Channels or Direct Messages category based on the type of channel. New DM and GM channels are
 // added to the Direct Messages category on each team.
-export function addChannelToInitialCategory(channel: Channel): ActionFunc {
+//
+// Unless setOnServer is true, this only affects the categories on this client. If it is set to true, this updates
+// categories on the server too.
```

Comment:
out of curiosity - what's the use case for setOnServer=false?

Reply:
It's for when adding a channel to Channels is a side effect of a server-side action, like when a custom category is deleted and all of its channels get moved back to the Channels/DMs category. When we delete a category, the server just deletes it and then returns that the category was deleted successfully, but on the client, we need to move the now-orphaned channels in the category to where they belong.